homeassistant/components/rainmachine/__init__.py

```diff
@@ -271,30 +263,77 @@
 class RainMachine:
     """Define a generic RainMachine object."""
 
-    def __init__(self, client, default_zone_runtime):
+    def __init__(self, hass, client, default_zone_runtime, scan_interval):
         """Initialize."""
+        self._async_unsub_dispatcher_connect = None
+        self._scan_interval_seconds = scan_interval
         self.client = client
         self.data = {}
         self.default_zone_runtime = default_zone_runtime
         self.device_mac = self.client.mac
+        self.hass = hass
+
+        self._api_category_count = {
+            PROVISION_SETTINGS: 0,
+            RESTRICTIONS_CURRENT: 0,
+            RESTRICTIONS_UNIVERSAL: 0,
+        }
+
+    async def _async_fetch_from_api(self, api_category):
+        """Execute the appropriate coroutine to fetch particular data from the API."""
+        if api_category == PROVISION_SETTINGS:
+            data = await self.client.provisioning.settings()
+        elif api_category == RESTRICTIONS_CURRENT:
+            data = await self.client.restrictions.current()
+        elif api_category == RESTRICTIONS_UNIVERSAL:
+            data = await self.client.restrictions.universal()
+
+        self.data[api_category] = data
+
+    @callback
+    def async_deregister_api_interest(self, api_category):
+        """Decrement the number of entities with data needs from an API category."""
+        # If this deregistration should leave us with no registration at all, remove the
+        # time interval:
+        if sum(self._api_category_count.values()) == 0:
+            if self._async_unsub_dispatcher_connect:
+                self._async_unsub_dispatcher_connect()
+                self._async_unsub_dispatcher_connect = None
+            return
+        self._api_category_count[api_category] += 1
+
+    async def async_register_api_interest(self, api_category):
+        """Increment the number of entities with data needs from an API category."""
+        # If this is the first registration we have, start a time interval:
+        if not self._async_unsub_dispatcher_connect:
+            self._async_unsub_dispatcher_connect = async_track_time_interval(
+                self.hass,
+                self.async_update,
+                timedelta(seconds=self._scan_interval_seconds),
+            )
+
+        self._api_category_count[api_category] += 1
+
+        # If the data hasn't been fetched for a particular category yet, do it:
+        if api_category not in self.data:
+            await self._async_fetch_from_api(api_category)
```

Comment:
This will cause at startup to fetch the same API endpoint multiple times. That's why Ring stored the task.

Reply:
I did something similar to you, which works, but I'm wondering if an `asyncio.Lock` would be a better construct?

Reply:
0a95107c066337c8b75149ce42f7bacfbbbd7d6c is what I'm thinking.